mantle/cmd/kola/qemuexec.go

```diff
@@ -64,89 +65,75 @@ func init() {
+	cmdQemuExec.Flags().StringArrayVar(&bindro, "bind-ro", nil, "Mount readonly via 9pfs a host directory (use --bind-ro=/path/to/host,/var/mnt/guest")
+	cmdQemuExec.Flags().StringArrayVar(&bindrw, "bind-rw", nil, "Same as above, but writable")
 	cmdQemuExec.Flags().BoolVarP(&forceConfigInjection, "inject-ignition", "", false, "Force injecting Ignition config using guestfs")
 }
 
-func renderFragments() (string, error) {
-	buf, err := ioutil.ReadFile(ignition)
-	if err != nil {
-		return "", err
-	}
-	config, _, err := v3.Parse(buf)
+func renderFragments(config v3types.Config) (*v3types.Config, error) {
 	for _, fragtype := range ignitionFragments {
 		switch fragtype {
 		case "autologin":
 			newconf := v3.Merge(config, conf.GetAutologin())
 			config = newconf
 			break
 		default:
-			return "", fmt.Errorf("Unknown fragment: %s", fragtype)
+			return nil, fmt.Errorf("Unknown fragment: %s", fragtype)
 		}
 	}
+	return &config, nil
+}
 
-	newbuf, err := json.Marshal(config)
-	if err != nil {
-		return "", err
-	}
-	tmpf, err := ioutil.TempFile("", "qemuexec-ign")
-	if err != nil {
-		return "", err
-	}
-	if _, err := tmpf.Write(newbuf); err != nil {
-		return "", err
-	}
-
-	return tmpf.Name(), nil
+func fs9pId(i int) string {
+	return fmt.Sprintf("fs%d", i)
 }
 
 func runQemuExec(cmd *cobra.Command, args []string) error {
 	var err error
-	cleanupIgnition := false
+	buf, err := ioutil.ReadFile(ignition)
+	if err != nil {
+		return err
+	}
+	config, _, err := v3.Parse(buf)
+	if err != nil {
+		return errors.Wrapf(err, "parsing %s", ignition)
+	}
 	if len(ignitionFragments) > 0 {
-		newconfig, err := renderFragments()
+		newconfig, err := renderFragments(config)
 		if err != nil {
 			return errors.Wrapf(err, "rendering fragments")
 		}
-		ignition = newconfig
-		cleanupIgnition = true
+		config = *newconfig
 	}
-	if kola.Options.IgnitionVersion == "v2" {
-		buf, err := ioutil.ReadFile(ignition)
-		if err != nil {
-			return err
-		}
-		config, _, err := v3.Parse(buf)
-		if err != nil {
-			return errors.Wrapf(err, "parsing %s", ignition)
-		}
-		ignc2, err := ignconverter.Translate3to2(config)
-		if err != nil {
-			return err
-		}
-		ignc2buf, err := json.Marshal(ignc2)
-		if err != nil {
-			return err
-		}
-		tmpf, err := ioutil.TempFile("", "qemuexec-ign-conv")
-		if err != nil {
-			return err
-		}
-		if _, err := tmpf.Write(ignc2buf); err != nil {
-			return err
-		}
-		if cleanupIgnition {
-			os.Remove(ignition)
-		}
-		cleanupIgnition = true
-		ignition = tmpf.Name()
+	for _, b := range bindro {
+		s := strings.Split(b, ",")
+		src := s[0]
+		dest := s[1]
+		config = v3.Merge(config, conf.Mount9p(src, dest, true))
+	}
+	for _, b := range bindrw {
+		s := strings.Split(b, ",")
+		src := s[0]
+		dest := s[1]
+		config = v3.Merge(config, conf.Mount9p(src, dest, false))
+	}
+	builder, err := platform.NewBuilderFromParsed(config, kola.Options.IgnitionVersion == "v2")
```

Comment:
This is an interesting one because what we really want to do is build up the qemu and Ignition config *simultaneously*; the looping above and below is a workaround.